### ISSUE TITLE
Write security proofs for nonstandard primitives

### ISSUE THREAD

**Issue by defuse** [zcash/zcash#792]
Looking over the current `protocol.pdf`, we are using the following nonstandard primitives. We should prove all of them secure using reasonable assumptions about SHA256.
- The four PRFs.
- The fixed-depth merkle tree which uses the SHA256 compression function as the hash.
- The way the coin commitment is computed.
- The way h_{sig} is computed.
- The way the key-private encryption works, which includes a KDF based on SHA256.

[Edit by Daira: the KDF is now based on BLAKE2b.]

**Comment by defuse** [comment]
Note that it's more valuable to prove that all of these are secure simultaneously, when used in the same or related protocols, than it is to prove that each is secure on its own.

**Comment by defuse** [comment]
I think this is an important exercise to avoid missing other InternalH-like attacks.

**Comment by matthewdgreen** [comment]
Replacing and/or proving something about the bottom two is probably a good exercise.

**Comment by zookozcash** [comment]
+1

Sigh, this makes me miss BLAKE2, which has a higher-level API so that things like domain separation and KDF are standard and widely analyzed. https://github.com/zcash/zcash/issues/706#issuecomment-187807410

But, alas, we have to keep building these things on top of the lower-level SHA256, and analyzing them ourselves, for now and then hopefully we can switch to BLAKE2 in the future.

**Comment by defuse** [comment]
For the bottom two points, we want to prove the same thing (after https://github.com/zcash/zips/issues/25 is done), which is that for any fixed number of bytes b and 1-byte constant c, SHA256(c || k || x) is a PRF where k is the 256-bit seed and x is the b-byte input.

I think that's equivalent to the following two assumptions:

1. SHA256compress(StandardIV, c || k || m), where k is a 32 byte seed and m is a 31 byte input and StandardIV is the standard SHA256 IV, is a PRF.
2. SHA256compress(k, m), where k is a 32 byte seed and m is a 64 byte input, is a PRF.
3. SHA256compress(k, m) is collision resistant for random k.

We can ignore padding by assuming it's part of the input x, since we're proving PRFness within a fixed b, and within a fixed b, pad(x) = x iff x = x.

If you draw out the Merkle–Damgård diagram, you can kind of see why 1,2,3 are probably enough: it's basically one PRF using another PRF's output as the seed... except there's a weird quirk where the leading part of x is "encoded" in the seed that gets passed along, so you also need assumption 3 to guarantee the second PRF "sees" changes to the leading part(s) of x.

I'm pretty confident about 1 and 3 but 2 is questionable, and I think 2 is *necessary* to prove the claim we want: If 2 is false then I can break the overall construction. As long as enough of x's trailing bits find themselves in the last block, I can use a distinguisher for SHA256compress(k, m) to distinguish the overall construction by putting its choices of m into the trailing bits of my choices of x and forwarding back the result I get.

I tried to get rid of assumption 2 and I ended up re-inventing HMAC. So: Why don't we just use HMAC? I think it's already been proven to be a PRF given 1 and 3. Neither KDF nor the computation of h<sub>sig</sub> are done in the circuit so I don't think there's any obstacle to using it.

**Comment by defuse** [comment]
[This paper](http://cseweb.ucsd.edu/~mihir/papers/hmac-new.html) claims in its abstract that [this paper](https://cseweb.ucsd.edu/~mihir/papers/kmd5.pdf) proved HMAC is a PRF, but the second paper actually proves unforgability under adaptive chosen message attacks, which doesn't imply PRFness (appending a 0 byte to the output of a MAC is still a MAC but definitely isn't a PRF). Maybe it's obvious to a cryptographer how the proof can carries over. In any case, if the first paper's abstract is to be believed they have even better proofs of PRFness than the second paper.

**Comment by defuse** [comment]
The coin commitment is secure in the random oracle model since it's essentially the same as the construction that's proved secure in [this midterm exam solution](http://www.cs.berkeley.edu/~daw/teaching/cs276-s06/mtsol.ps).

I don't like that definition of (t, ε)-concealing because it doesn't account for the possibility of the adversary getting more than one commitment to the same value. In fact if we consider that then our scheme only has 96 bits of security since if I commit to the same a<sub>pk</sub>, v, and ρ 2^96 times then the attacker only has to try 2^96 guesses for r before they match one of my commitments, because r is only 192 bits (you can trade-off number of commitments for attack runtime).

Let's not worry about that because it's impractical and we don't have to worry about it anyway because the faerie gold fix makes it super unlikely ρ will ever repeat a single time, let alone 2^96 times.

**Comment by defuse** [comment]
The four PRFs are obviously independent PRFs assuming the SHA256 compression function is a PRF when the seed bits are put in places we're putting the seed bits in and all other bits are considered part of the input. Whether or not that's true is a question of empirical SHA256 cryptanalysis. I'm sure if evidence that it were false surfaced the community would question the security of SHA256.

**Comment by defuse** [comment]
We *do* need to take a closer look at the Merkle tree. That's being tracked in #779.

**Comment by defuse** [comment]
I'm now satisfied that after #779 and https://github.com/zcash/zips/issues/25 are resolved then we'll be able to point to proofs of security for all our nonstandard primitives, so I think this ticket can be closed.

**Comment by defuse** [comment]
Whoops, this definitely can't be closed. I only looked at KDF but nothing else about how the encryption works. I have no idea how to prove that secure so I'll leave it up to @daira and @matthewdgreen.

**Comment by daira** [comment]
For the KDF, djb says in the main [Curve25519 paper](https://cr.yp.to/ecdh/curve25519-20060209.pdf):
> **Choice of key-derivation function.** There are no theorems guaranteeing the safety of any particular key-derivation function H with, e.g., 512-bit output. Some silly choices of H are breakable. As an extreme example, if H outputs just 64 bits followed by all zeros, then an attacker can perform a brute-force search for those 64 bits.
On the other hand, from the perspective of a secret-key cryptographer, it seems very easy to design a safe function H. A small amount of mixing, far less than necessary to make a safe secret-key cipher, stops all known attacks.

This suggests to me that SHA-256 is sufficient. I'm not wedded to this choice if someone can make an argument that something else is preferable, although I do think we should keep things simple unless there's a good reason for added complexity.

**Comment by daira** [comment]
AEAD_CHACHA20_POLY1305 has a proof of security as an AEAD which is referenced in the RFC. (The tweak I made to reduce the number of ChaCha20 blocks does not affect the proof, but I'll consider removing that to make it more standard if it doesn't affect forward compatibility with the viewing keys proposal.)

**Comment by daira** [comment]
There's been a lot of discussion about this over on https://github.com/zcash/zips/issues/25 . To summarise, for the KDF I now propose

  SHA-512((2-bit i) || trunc<sub>246</sub>(h<sub>Sig</sub>) || dhsecret || epk || pk<sup>new</sup><sub>enc,i</sub>)

This choice is supported by section 7 of [Computational Extractors and Pseudorandomness](https://eprint.iacr.org/2011/708). The inclusion of trunc<sub>246</sub>(h<sub>Sig</sub>) in the KDF input helps because:

* it can be treated as a public uniform-random seed since it's the output of another hash, and is independent of the other KDF inputs. Such a value is necessary for the section 7 proof, and it is long enough for the proof to give practical guarantees for the parameters we are using and with reasonable assumptions on SHA-512;
* it is authenticated by the Pour proof as having been chosen with knowledge of a<sup>old</sup><sub>sk,i</sub>, which is a bonus since that prevents an adversary from changing it in a chosen-ciphertext attack. (Note however that if an attacker were to break this authentication by breaking the SNARK for instance, then this would still be a good heuristically secure KDF.)

**Comment by daira** [comment]
We'll use the standard AEAD_CHACHA20_POLY1305. (I looked at how much that would cost us if we later do the full-featured viewing keys with circuit enforcement, and it would only cost one extra ChaCha20 block relative to my original design.)

**Comment by daira** [comment]
I'm happy about the security of the PRF and KDF now, but less so about the security of the commitment. Does anyone know of a proof that a Merkle-Damgård hash makes a secure (computationally binding and computationally hiding) commitment scheme given reasonable assumptions about the hash other than modeling it as a random oracle? (The proof that @defuse [mentioned](https://github.com/zcash/zcash/issues/792#issuecomment-197512482) models it such, and as he mentioned, has some deficiencies in its definitions.]

I can only find this: [Practical and Provably-Secure Commitment Schemes from Collision-Free Hashing](https://crypto.cs.mcgill.ca/~crepeau/COMP647/2007/TOPIC07/Halevi-Micali.pdf) which explains in section 1.3 why collision resistance is not sufficient to prove security of the scheme we're considering. I didn't expect it to be, but it's unclear precisely what property of the hash is needed, which makes me uneasy. The scheme they suggest instead is undesirably heavy for the SNARK, since it requires a UOWHF as well as two CRHF invocations. (Hmm, perhaps we could use Poly1305 for the UOWHF? That's very SNARK-efficient, but I'd rather avoid implementing more primitives for the SNARK if at all possible.)

**Comment by daira** [comment]
Also we need to decide what length is needed for r.

**Comment by matthewdgreen** [comment]
If you treat the hash as a PRF where the commitment randomness is the seed for the PRF you get computational hiding in the obvious way. If you treat the hash as collision resistant, you get computational blinding in the obvious way. 

Proving those properties of MD constructions requires various well-known assumptions about the compression function. There should be plenty of literature on that.

**Comment by matthewdgreen** [comment]
Sorry, blinding->binding. All the cited paper points out is that CR is not sufficient for a secure commitment. CR+PRF is.

**Comment by daira** [comment]
The question I was getting at is whether the Merkle-Damgård structure causes problems. I see how to approach this now.

We have COMM<sub>r</sub>(x || y) = COMM'<sub>r</sub>(H(x) || y), where
* x is the first block of input
* y is the second block
* H is SHA256Compress (with standard IV)
* COMM' applies SHA256Compress with the first 256 bits of its input in the chaining variable, and the rest of its input and the seed r in the input block.

Suppose we assume COMM' to be a computationally binding and hiding commitment. If H is collision-resistant [*], then COMM is as secure for binding as COMM'. (Any break of the binding property on COMM requires an adversary to either break binding for COMM' or explicitly find a collision for H.) Also COMM is as secure for hiding as COMM' for **any** H.

This rules out the Merkle-Damgård structure being a problem, *provided* (for this particular security argument) that r is fully in the second block. In our instantiation, r is indeed fully in the second block.

Furthermore, H is domain-separated from other uses of SHA256Compress. COMM' is not, although the IV is almost surely not the standard one, so that probably doesn't matter. It would be nice to tweak the encoding of arguments so that COMM' is domain-separated.

We still need to consider the length of r.

[*] Note that if SHA256Compress with standard IV is not collision-resistant, then SHA-256 is not collision-resistant for a 2-block input.

**Comment by matthewdgreen** [comment]
My point was to list the requirements you need from the overarching hash construction. Given those requirements you can determine whether the MD construction provides them.

The notation you give is confusing. However, the main comment is that you need an additional assumption to prove that the construction is hiding. I don’t understand what "Also COMM is as secure for hiding as COMM' for *any** H.” means. Is that an assumption or a conclusion? Are you saying that H(), seeded on “x” and evaluated on the standard IV, is a PRF?

Rather than trying to prove this from scratch using parts, have you considered using HMAC — which has about the same performance and also includes well-studied security proofs showing that it is a PRF and collision resistant (under standard assumptions about the compression function)?

Matt

**Comment by daira** [comment]
No additional assumption is needed for hiding. H is not assumed to be a PRF. It is obvious that if COMM' hides H(x) and y, then it hides x and y, because the adversary gets no other information about x. This would be true even if H were a very weak hash, or the identity function.

The proofs about HMAC (for example, in https://eprint.iacr.org/2006/043) assume that the compression function is a PRF. Since we already assume that PRF<sup>sn</sup> and PRF<sup>ρ</sup> are collision-resistant PRFs, there doesn't seem to be any value gained from using HMAC in terms of what we need to assume about SHA-256-compress. Note that HMAC with an input key (K<sub>out</sub> || K<sub>in</sub>) of 512 bits, and an input a<sub>pk</sub> || v || ρ of 72 bytes, requires 5 compression function evaluations instead of 2. That would be 12 extra hashes in the circuit.

**Comment by daira** [comment]
I wrote:
> Furthermore, H is domain-separated from other uses of SHA256Compress. COMM' is not, although the IV is almost surely not the standard one, so that probably doesn't matter. It would be nice to tweak the encoding of arguments so that COMM' is domain-separated.

It definitely doesn't matter. The lack of domain separation between COMM' and the other uses of SHA256Compress would only be an issue if the IV in a particular case were the same as the standard IV. But that could only happen for its usage in the instantiation of COMM if SHA256Compress(StandardIV, x) = StandardIV. Then we would have SHA-256(x || M) = SHA-256(M) for any M, contradicting the assumed collision resistance of SHA-256.

**Comment by daira** [comment]
On the length of r: we're aiming for at least a 128-bit security level. Use of longer cryptovalues (e.g. for a<sub>sk</sub>, φ, etc.) is justified because we want to ensure that an attacker expending 2<sup>128</sup> work has *negligable* success probability even when the adversary has multiple targets resulting from the protocol being run many times (see Dan Bernstein's excellent paper [Understanding brute force](https://cr.yp.to/snuffle/bruteforce-20050425.pdf)). However, as @defuse [pointed out](https://github.com/zcash/zcash/issues/792#issuecomment-197512482), the only way this can cause a problem is if the input to COMM' repeats. This could only happen if there is a collision on H or if ρ repeats, and ρ can only repeat if there is a collision on PRF<sup>ρ</sup> or if φ repeats. φ is chosen randomly and is 252 bits. (Note that part of ρ is in the first input block and part in the second, but this doesn't matter.)

Therefore, 192 bits is sufficient for r because the adversary is prevented from performing a multi-target attack, and has at most a 2<sup>-64</sup> chance of correctly guessing r after expending 2<sup>128</sup> work.

**Comment by defuse** [comment]
Classically 192 bits is sufficient but not if your adversary has a quantum computer: https://github.com/zcash/zcash/issues/805#issuecomment-202508564 (I personally think launching with 192 bits is fine).

**Comment by daira** [comment]
That's a good point @defuse. Let's increase r to 256 bits. All other cryptovalues are at least 252 bits, so I think this suffices for 126-bit post-quantum privacy when addresses are secret. (For post-quantum privacy when using non-secret addresses, we'd have to change the public key encryption as stated in #805.)

**Comment by defuse** [comment]
+1 to the reasoning in https://github.com/zcash/zcash/issues/792#issuecomment-202046411

**Comment by matthewdgreen** [comment]
@daira said: No additional assumption is needed for hiding. H is not assumed to be a PRF. It is obvious that if COMM' hides H(x) and y, then COMM hides x and y, because the adversary gets no other information about x. This would be true even if H were a very weak hash, or the identity function."

This is not correct. One of the two inputs to COMM' must be random in order for COMM' to have security properties as a standard commitment scheme. You'll notice we use the notation COMM_r(m) in the paper to make this clear. You've moved to an alternative notation, which is why I asked for clarification.

Here's why your claim is not correct. Suppose COMM' is a secure hiding/binding commitment scheme if the first block is random, but holds neither property if the first block has some particular structure. If we assume H() is an arbitrary hash function then it would be easy to contrive H() for which the outputs _always_ have this structure, and thus the construction would always be insecure.

Conversations like these are making me more and more uncomfortable about your decision to design custom protocols. I would be much more comfortable using HMAC instead, or just assuming random oracles, rather than assuming some bogus security properties that you cannot justify from these analyses.

**Comment by defuse** [comment]
@matthewdgreen: You may have missed that COMM' takes a random r as an argument; Daira wrote "We have COMMr(x || y) = COMM'<sub>r</sub>(H(x) || y), where." I think the confusion is because when @daira referenced the function by name she omitted the subscript (since it's an argument, not part of the name) which is different from the notation in the Zerocash paper.

**Comment by daira** [comment]
@defuse wrote:
> I think the confusion is because when @daira referenced the function by name she omitted the subscript (since it's an argument, not part of the name) which is different from the notation in the Zerocash paper.

Just a nitpick, but this is exactly the same usage as in the Zerocash paper (e.g. see step 1 in section 1.3).

**Comment by daira** [comment]
@matthewdgreen wrote:
> Suppose COMM' is a secure hiding/binding commitment scheme if the first block is random, but holds neither property if the first block has some particular structure. If we assume H() is an arbitrary hash function then it would be easy to contrive H() for which the outputs always have this structure, and thus the construction would always be insecure.

That's true and also irrelevant. In that case COMM' is not a secure hiding/binding commitment scheme, contrary to assumption. There's also no reason to believe, given the structure of the SHA-256 compression function, that it would have such a weakness.

**Comment by matthewdgreen** [comment]
You posed a question about building commitments from hash functions. Specifically you asked about assumptions needed to build hiding commitments from these functions. When building hiding commitments from hash functions the absolute most important thing -- from a formal analysis perspective -- is where the randomness is being inserted, and what assumptions we're making about the compression function (is it a PRF, etc.). Those were the questions you started with, and I pointed out that it depends fundamentally on those details. 

You then switched to a different point about commitments that omitted the critical element of commitment randomness. From my point of view this question made no sense in context and I asked what was going on with the notation. It made /particularly/ no sense, since we were having an intricate discussion about building hiding commitment schemes from hash functions, and suddenly proposing a commitment scheme and omitting the randomness from your construction made this issue even more confusing.

Rather than trying to clarify the notation so I could answer your question, you went forward with some statement about "obviously" this is true. If your point is that given a commitment scheme (with some randomness not specified) then committing to H(x) for *some collision resistant hash function H()* rather than x itself will also give a secure commitment scheme, that is true. But you did not ask that question, you asked something entirely different -- and did it in a weird, confusing context.

If you want to build new protocols and conduct formal analysis of the details of MD hash functions -- rather than assuming random oracles -- I am ok with doing that. But formal analysis requires precision and exact notation, and what you are doing is not that. At this point I don't even understand how the second question about commitments relates to the initial question about commitments using MD hash functions, and I don't know what the construction we currently have is. I know that HMAC meets all of our requirements, but whatever we're using -- no idea at all, none, and quite frankly I tend to assume things are insecure unless they've been properly analyzed.

As far as Zerocash goes, you may be right that there are typos in the Zerocash paper -- and IF THOSE COME UP WE SHOULD FIX THEM -- not use them as justification for making further mistakes in analyzing this new scheme. Please don't do this. And if it becomes obvious that we are not communicating correctly, take the time to EXPLAIN what you are doing so instead of "winning arguments" we are all working towards analyzing the security of the same scheme, together.

**Comment by daira** [comment]
@matthewdgreen All of your criticisms seem to be due to you misreading my comment, which clearly did include the randomness and was a correct security proof. It's fine to assume things are "insecure unless they've been properly analyzed", but what I gave was a proper analysis, with precise and exact notation. And what I said was obvious was, in fact, obvious.

HMAC is 2.5 times more expensive.

I have never used errors or imprecisions in the Zerocash paper to justify anything.

**Comment by matthewdgreen** [comment]
Thanks Taylor, Daira. I'm sorry for my ranty rant earlier. Seeing the full specification is very helpful. I think I misunderstood the implication of Daira's comment earlier because (1) I didn't expect to see randomness added at the end of an MD hash construction, and (2) I was on a mobile phone and the fonts were small. I really appreciate you writing out the full thing.

My one concern about the H( H(m) || r) type constructions is that -- done in a specific way -- it might allow for something like a double spending attack if m contains only the serial number randomness and there is an arbitrary collision in the SHA2 compression function. This is a pretty rarified concern from the start, and it is not relevant to the current construction.

**Comment by daira** [comment]
@matthewdgreen Apology accepted.

> My one concern about the H( H(m) || r) type constructions is that -- done in a specific way -- it might allow for something like a double spending attack if m contains only the serial number randomness and there is an arbitrary collision in the SHA2 compression function. This is a pretty rarified concern from the start, and it is not relevant to the current construction.

I want to understand this concern fully &mdash; I was under the impression that the protocol as a whole couldn't avoid being vulnerable to arbitrary hash collisions? For example, there is a balance violation attack if the Merkle tree hash allows collisions, regardless of how the commitment is instantiated (I know @zookozcash looked at whether there was a way to avoid that by using SPR trees [DOTV2008] or similar, and concluded that they weren't applicable). It's true that we don't want to rely on unnecessarily strong properties in any particular construction.

[DOTV2008] Erik Dahmen, Katsuyuki OKEYA, Tsuyoshi TAKAGI, and Camille Vuillaume,
"Digital Signatures Out of Second-Preimage Resistant Hash Functions,"
In Post-Quantum Cryptography, Lecture Notes in Computer Science Vol. 5299, pp. 109-123, Springer-Verlag 2008.

**Comment by defuse** [comment]
I thought of a better game to define the hiding property, and I'm writing it in this ticket so that I can erase it from my whiteboard, since I want to actually try and write a full proof for our construction on a fun Friday or something:

**Strong Computational Hiding:**

1. Alice selects a random bit b.
2. Adversary (adaptively) sends Alice N pairs (m<sub>0</sub>, m<sub>1</sub>) of her choice, and Alice replies with Commit<sub>r</sub>(m<sub>b</sub>) for a random r.
3. Adversary outputs a guess for b.

It's similar to IND-CPA but without the key. I think it captures all the properties we want since r is never supposed to be reused and if the Adversary knew r they'd know all the inputs and they could just compute the commitment for themselves. Alice never reveals any of the commitments, but that doesn't matter because the Adversary can get a commitment to a message M of their choice by sending (M, M).

**Comment by matthewdgreen** [comment]
Why N pairs?

**Comment by defuse** [comment]
@matthewdgreen: By N pairs I meant one pair at a time, seeing the response after each one, up to however many queries the adversary is allowed to ask. I'm not sure what the standard way to describe that is (it's probably a property of the adversary not of the game like what I wrote suggests), sorry if that's confusing.

**Comment by defuse** [comment]
**Fun Friday Stuff** 

Daira's reasoning tells us our commitment scheme is secure only if we assume SHA256Compress is a secure commitment scheme (under whichever definition you want to use). So in total, we'll be assuming, for the Zcash protocol,

1. SHA256Compress is collision resistant.
2. SHA256Compress is a PRF.
3. SHA256Compress is a hiding commitment scheme.

I tried to eliminate assumption 3 by showing that SHA256Compress is Strong Computational Hiding under the assumption that it's a PRF. I couldn't do it, but I got close. It's Strong Computational Hiding under an assumption which is (probably) stronger than assuming it's a PRF but (probably) still weaker than assuming it's a random oracle. [Here are my notes](https://gist.github.com/defuse/c525cfaaebfb5e63346bfd36c077d7b6)

**Comment by matthewdgreen** [comment]
Good thinking. But, what if the N-pair element of the game is unnecessary? 

For example: imagine that you have an adversary that succeeds against the N-pair "strong hiding" game with non-negligible advantage. See if you can construct an adversary that succeeds with non-negligible advantage against a 1-pair "[not-strong] hiding" game. Then see what that does to your assumption (3).

**Comment by daira** [comment]
@matthewdgreen [said above](https://github.com/zcash/zcash/issues/792#issuecomment-201911845) that CR+PRF is sufficient for a secure commitment. Matt, can you link us to a proof of that? (It certainly seems plausible.)

**Comment by daira** [comment]
@defuse There's some imprecision in your assumptions:
1. SHA256Compress(StandardIV, sep || input) is collision-resistant
2. SHA256Compress(StandardIV, sep' || key || input) is a PRF
3. SHA256Compress(input_x, sep'' || key || input_y) is a secure commitment to input_x || input_y

**Comment by daira** [comment]
@defuse wrote:
> We *do* need to take a closer look at the Merkle tree. That's being tracked in #779.

#779 only covers the H(0) = 0 issue. There's also the fact that the use of the compression function in the Merkle tree isn't domain-separated from other uses (#501, which I think may have been closed prematurely). I thought of an efficient way to fix that: use a different IV for each layer of the tree. That is, each hash becomes SHA256Compress(SHA256Compress(sep || layer), left || right). Since the IV is constant for each layer, it can be inlined into the circuit. This should be collision-resistant by a variation of the standard Merkle-Damgård collision-resistance proof (without padding, but for a fixed length of 2 blocks). It should also address #404.

(Note that we already need to have a circuit implementation of SHA256Compress for arbitrary IV in order to implement the new commitment scheme.)

**Comment by ebfull** [comment]
* The SHA256 compression function is already designed to operate over arbitrary IV.
* Modifying the tree construction to remove the fixpoint (probably?) doesn't require modifying the circuit.
* Modifying the circuit to support domain separation for the merkle tree is not necessary for security in my mind, hence why #501 was closed. In fact, there is probably more risk in screwing up adding domain separation than leaving it alone. Honestly, it is beginning to be too late to add things like this to the spec unless there's a clear security danger.

**Comment by ebfull** [comment]
The fixpoint is also what allows us to say the initial root is null, an invariant we expect throughout our code.

**Comment by defuse** [comment]
**Continuation of Fun Friday Stuff**

@matthewdgreen: Let's call the security property with the game modified to one query "Weak Computational Hiding." The reason I think Weak is insufficient and Strong is better is because Strong allows collisions in the randomness to break the scheme. For example if the randomness is 128 bits, then you can have 128 bits of security in the Weak Computational Hiding model, but only 64 bits in the Strong Computational Hiding model: Send 2^64 queries with ("0", "1") for the first half of your queries and ("0", "0") for the second half, if you get a collision between the first and second halves of the responses output 0 otherwise output 1; you're right with high probability.

I found an *optimal* way to turn an arbitrary adversary for the Strong game into an adversary for the Weak game. The reduction is optimal in the sense that there's no [fully black-box reduction](https://eprint.iacr.org/2013/101.pdf) that gives an advantage that's significantly better (by a factor of 4 or more). The details are in the "Is the many-query game necessary, or would a one-query game suffice?" section of [my notes](https://gist.github.com/defuse/c525cfaaebfb5e63346bfd36c077d7b6). TL;DR: "Removing" any single one of the Strong game adversary's query pairs decreases their advantage by some amount Δ; the advantage you get in the weak game is Δ/2 (maximized over which query pair you chose to "remove"), and the existence of a better reduction would imply a contradiction.

I wonder if it's worth finishing the proof and writing it up... The result is a tight upper and lower bound on the complexity of reducing one problem to another and it seems like the proof technique could be reused in other "many queries" vs. "one query" situations.

Conclusions that are useful for *this* ticket: I *don't* think CR + PRF implies Strong Computational Hiding, even though it might (probably does!) imply Weak Computational Hiding, which is really all we need.

**Comment by daira** [comment]
@defuse Since there's no key, can't the adversary compute Commit<sub>r</sub>(m<sub>b</sub>) for all queries except the last on their own?

**Comment by defuse** [comment]
CR+PRF does imply Weak Computational Hiding. How CR implies binding is obvious (I think). For hiding, suppose PRF<sub>r</sub>(m) is a PRF and that A is an adversary that has advantage ADV in the Weak Computational Hiding game with PRF. Construct a distinguisher for the PRF as follows:

1. Select a random bit b.
2. Run A, responding to (m<sub>0</sub>, m<sub>1</sub>) with PRF<sub>k</sub>(m<sub>b</sub>) using just one call to the PRF<sub>k</sub> oracle.
3. Let b' be A's output bit. If b' == b then output 1. Otherwise output 0.

In the real world this adversary outputs 1 with probability ADV + 1/2 since it's equivalent to the Weak game and by the definition of advantage in the Weak game. In the random world this adversary outputs 1 with probability 1/2 since all A is given is a random string so A's choice can't be correlated to the b we picked. So the advantage is (ADV + 1/2) - 1/2 = ADV.

**Comment by defuse** [comment]
@daira: The inner adversary's queries are pairs (m<sub>0</sub>, m<sub>1</sub>). Since there's no key the outer adversary can respond with *either* Commit<sub>r</sub>(m<sub>0</sub>) or Commit<sub>r</sub>(m<sub>0</sub>). Computing Commit<sub>r</sub>(m<sub>b</sub>) theirselves presumes they know b (in which case they've already won).

**Comment by daira** [comment]
Oh, r *is* a key, because it's only revealed on a decommit. I see why adaptive queries are necessary now; carry on :-)

**Comment by matthewdgreen** [comment]
Consider a slight reformulation of @defuse's multi-query game, as follows. Let Exp0 be the multi-query experiment run with b=0, and let Exp1 be the multi-query experiment run with b=1. One formulation of the security property is that for all p.p.t. adversaries A, the value \epsilon = |Pr[A outputs "1" in Exp1] - Pr[A outputs "1" in Exp0]| should be negligible.

Now consider an adversary A such that \epsilon is non-negligible. It should be easy to show that there exists an adversary B that succeeds with related probability against the single-query game. In fact, this is similar to what @defuse already did -- minus the requirement that A output its queries in advance.

The full proof is dull, but the rough idea is to use a hybrid argument. You start with Exp0 (where all challenges are answered with b=0) then for i=1 to q you consider a series of q intermediate "games" where at each game the bit used to answer the i^[th} query is switched from 0 to 1. You eventually end up with Exp1. The key is to note that if \epsilon (the difference between Exp0 and Exp1) is non-negligible, then there must exist at least one pair of consecutive intermediate games where the probability of distinguishing the two games is also non-negligible -- and then you use this adversary to construct a second adversary against the 1-query definition.

For a much more formal version of all this, see http://cseweb.ucsd.edu/~mihir/cse207/w-asym.pdf, section 11.3. That proof deals with IND-CPA encryption, but the same logic applies here. 

I am trying to respond more thoughtfully to @defuse's arguments about randomness collisions. I think these are already captured in the 1-query game. Think of it this way: even in the 1-query game, the _adversary can already simulate the commitment challenge oracle itself_ and create randomness collisions. Thus she gets no additional power from having an oracle that provides her with many challenges. Formulating this concretely is a pain though.

**Comment by defuse** [comment]
@matthewdgreen Thanks for the awesome reply! 

I see how both "A many-query p.p.t. adversary implies a one-query p.p.t. adversary" *and* my randomness collisions argument can be true at the same time. My randomness collision argument *relies* on sending enough queries for there to likely be a randomness-collision in one of the oracle calls (which can't be simulated because of the unknown secret bit b). This makes the number of intermediate games "q" exponential, something like 2^(|r|/2), in which case the following would NOT be true: "there must exist at least one pair of consecutive intermediate games where the probability of distinguishing the two games is also non-negligible".

So the proof-sketch in my notes shows that reduction I gave the best you can do given an arbitrary adversary *which may not be constrained to polynomial time*. Once you add the assumption that the adversary is polynomial time you can use the argument in Matt's comment above.

**Comment by defuse** [comment]
I've had a lot of fun (and learned a lot) writing proofs for this ticket but I think it's getting off-topic and @matthewdgreen's and @daira's time are super valuable, so thank you both for considering my arguments! I'm now satisfied with the security of our note commitment function.

**Comment by daira** [comment]
This ticket concerns the primitives used in Sprout. All of the necessary questions were resolved, but before closing it I need to make sure that any non-obvious security arguments (for example the one about the use of SHA-256 in NoteCommit<sup>Sprout</sup>) are copied to the spec. In the case of Sapling, I was specifying needed security properties for each primitive as I went along (and hair-raising optimisations like the windowing in the Pedersen have explicit correctness proofs), so I think we're fine there.

**Comment by daira** [comment]
Hmm, I think that the argument for the security of NoteCommit^Sprout never did get copied to the protocol spec. (It should go in section 8.5.)

**Comment by daira** [comment]
> Hmm, I think that the argument for the security of NoteCommitSprout never did get copied to the protocol spec. (It should go in section 8.5.)

Fixed in protocol spec version 2022.2.19 (specifically https://github.com/zcash/zips/commit/4ef578706b2ce52dd29b681eeb36921729a67dcf).

**Comment by daira** [comment]
Just to tie up a few loose ends:

* If you want a conservatively secure PRF or CRPRF or KDF that does not need to be especially circuit-efficient for a new design, use a hash with a keyed mode like BLAKE2b/s or BLAKE3. The reason for using SHA-256 was that we already had an implementation of it, but not of BLAKE2s, in the circuit at the time. (BLAKE2s was added for CRH<sup>ivk</sup> in the Sapling circuit.) BLAKE2s and BLAKE3 are smaller than SHA-256 in a circuit.
* You can alternatively assume that SHA256Compress is free-start collision-resistant (i.e. collision-resistant on an input that also includes the chaining variable), in order to prove that NoteCommit<sup>Sprout</sup> is binding. The assumption that COMM' is binding is weaker; it's essentially the same problem but the adversary must use a random r rather than chosen r.
